compass/compass-cw/src/contract.rs

```diff
@@ -0,0 +1,156 @@
+            return Ok(());
+        }
+    }
+    bail!("Insufficient Power");
+}
+
+fn check_validator_signatures(deps: Deps, consensus: &Consensus, msg: &[u8]) -> Result<()> {
+    let valset = VALSET.load(deps.storage)?;
+    ensure!(
+        valset.valset_id == consensus.valset_id,
+        "Must use current valset"
+    );
+
+    let hash = digest(&SHA256, msg);
+    let mut cumulative_power: u64 = 0;
+    for (validator, &power, sig) in izip!(&valset.validators, &valset.powers, &consensus.signatures)
+    {
+        check_signature(deps, validator, hash.as_ref(), sig)?;
+        cumulative_power += power as u64;
+        if cumulative_power >= POWER_THRESHOLD {
+            return Ok(());
+        }
+    }
+    bail!("Insufficient Power");
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> Result<Response> {
+    check_validator_signatures(deps.as_ref(), &msg.consensus, &msg.payload)?;
+    let payload = serde_json::from_slice(&msg.payload)?;
+    let id = SMART_CONTRACT_ID.load(deps.storage)?;
+
+    use crate::msg::ExecutePayload::*;
+    match payload {
+        UpdateValset(new_valset, smart_contract_id) => {
+            ensure!(smart_contract_id == id, "Wrong compass instance");
+            update_valset(deps, env, info, &new_valset)
+        }
+        SubmitLogicCall(args, message_id, smart_contract_id, deadline) => {
+            ensure!(smart_contract_id == id, "Wrong compass instance");
+            submit_logic_call(deps, env, info, args, message_id, deadline)
+        }
+    }
+}
+
+/// This updates the valset by checking that the validators in the current valset have signed off on the
+/// new valset. The signatures supplied are the signatures of the current valset over the checkpoint hash
+/// generated from the new valset.
+/// Anyone can call this function, but they must supply valid signatures of constant_powerThreshold of the current valset over
+/// the new valset.
+/// valset: new validator set to update with
+/// consensus: current validator set and signatures
+fn update_valset(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    new_valset: &Valset,
+) -> Result<Response> {
+    let valset_id = VALSET_ID.load(deps.storage)?;
+    // check if new valset_id is greater than current valset_id
+    ensure!(new_valset.valset_id > valset_id, "Invalid Valset ID");
+    check_validator_power(&new_valset.powers)?;
+    VALSET.save(deps.storage, new_valset)?;
+    VALSET_ID.save(deps.storage, &new_valset.valset_id)?;
+    Ok(Response::new())
+}
+
+/// This makes calls to contracts that execute arbitrary logic
+/// message_id is to prevent replay attack and every message_id can be used only once
+fn submit_logic_call(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    args: LogicCallArgs,
+    message_id: Uint256,
+    deadline: u64,
+) -> Result<Response> {
+    ensure!(env.block.height < deadline, "Timeout");
```

Comment:
```suggestion
    ensure!(env.block.height < deadline, "Timeout");
```

deadline is time in UTC. Could you check for env.block.time ?

Reply:
Is it in seconds or millis? And would it be okay to indicate which in the name of the variable?